auth/src/main/java/org/aerogear/auth/impl/OIDCAuthCodeImpl.java

```diff
@@ -20,6 +43,111 @@ public OIDCAuthCodeImpl(final AuthServiceConfig config) {
+        if (userIdentity != null) {
+            // get the users username
+            if (userIdentity.has(USERNAME) && userIdentity.getString(USERNAME).length() > 0) {
+                username = userIdentity.getString(USERNAME);
+            }
+        }
+        return username;
+    }
+
+    private String getEmail() throws JSONException {
+        String emailAddress = "Unknown Email";
+        if (userIdentity != null) {
+            // get the users email
+            if (userIdentity.has(EMAIL) && userIdentity.getString(EMAIL).length() > 0) {
+                emailAddress = userIdentity.getString(EMAIL);
+            }
+        }
+        return emailAddress;
+    }
+
+    private List<IRole> getRoles() throws JSONException {
+        List<IRole> roles = new ArrayList();
+        if (userIdentity != null) {
+            List<RealmRole> realmRoles = getRealmRoles();
+            if (realmRoles != null) {
+                roles.addAll(realmRoles);
+            }
+            List<ClientRole> clientRoles = getClientRoles();
+            if (clientRoles != null) {
+                roles.addAll(clientRoles);
+            }
+        }
+        return roles;
+    }
+
+    private List<RealmRole> getRealmRoles() throws JSONException {
+        List realmRoles = new ArrayList<RealmRole>();
+        if (userIdentity.has(REALM) && userIdentity.getJSONObject(REALM).has(ROLES)) {
+            String tokenRealmRolesJSON = userIdentity.getJSONObject(REALM).getString(ROLES);
+
+            String realmRolesString = tokenRealmRolesJSON.substring(1, tokenRealmRolesJSON.length() - 1); //remove the array brackets
+            String roles[] = realmRolesString.split(COMMA);
+            realmRoles = Arrays.asList(roles);
+        }
+        return realmRoles;
+    }
+
+    private List<ClientRole> getClientRoles() throws JSONException {
+        List clientRoles = new ArrayList<ClientRole>();
+
+        AuthServiceConfig authConfig = this.getConfig();
+        JSONObject authJSON =  authConfig.toJSON();
+
+        if (authJSON.has(RESOURCE)) {
+            String initialClientID = authJSON.getJSONObject(RESOURCE).toString();
+
+            if (userIdentity.has(CLIENT) && userIdentity.getJSONObject(CLIENT).has(initialClientID)
+                    && userIdentity.getJSONObject(CLIENT).getJSONObject(initialClientID).has(ROLES)) {
+                String tokenClientRolesJSON = userIdentity.getJSONObject(CLIENT).getJSONObject(initialClientID).getString(ROLES);
+
+                String clientRolesString = tokenClientRolesJSON.substring(1, tokenClientRolesJSON.length() - 1);
+                String roles[] = clientRolesString.split(COMMA);
+                clientRoles = Arrays.asList(roles);
+            }
+        }
+        return clientRoles;
+    }
+
+    private JSONObject getIdentityInformation(final ICredential credential) throws JSONException {
+        String accessToken = credential.getAccessToken();
```

Comment:
This won't compile will it? ICredential doesn't declare a `getAccessToken` function

Reply:
Ya you're right.  There' another compile error too. Nice catch, thanks